### ISSUE TITLE
Terrible latency compared to `appengine` package

### ISSUE THREAD

**Issue by pjvds** [GoogleCloudPlatform/gcloud-golang#135]
## TLDR;

We are currently evaluating the appengine platform and we see terrible latency when getting data from the datastore.

The interesting part is that this only happens if we get the data with `cloud/datastore` package, when we do the same thing with the `appengine/datastore` package the latency problems are gone.

## Context

We are running our app in the appengine (not the compute engine). We serve a single HTTP endpoint that does a `GetMulti` call with 10 keys.

The code is identical, we only switched package.

The larger the concurrency, the bigger the difference becomes. The `appengine` package seems to scale pretty well while the `cloud` package implementation only gets worse.

## The numbers

**appengine** avg latency **82 ms**, detailed report: http://ldr.io/1S0JLN8
**cloud**: avg latency **2821 ms**, detailed report: http://ldr.io/1Hlzgkj

## Urlfetch

It looks like it has something to do with the `urlfetch.Fetch` that is used to do the actual request to the datastore. We create a context as shown [in the example](https://github.com/GoogleCloudPlatform/gcloud-golang/blob/master/examples/storage/appengine/app.go#L77). Traces show `urlfetch.Fetch` items with _bad_ latency.

**Comment by PSG-Luna** [comment]
cloud/datastore currently uses a transparent translator that translates down an API version. Could this be related to the warmup time of said translator ? Does the average latency improve at all after a period of consistent requests ?

**Comment by pjvds** [comment]
@PSG-Luna It does start terribly slow for the first few seconds, and then stabilize at slow (2s latency), even if we let it run for 15 minutes.

**Comment by pjvds** [comment]
What we have tried so far:

- run the same code in the **compute engine**
- run the same code in a new project in **Europe**
- run the same code in a new project in **US**
- increase warmup time to **30 minutes**

We still see the same terrible latency in all cases listed above, roughly **2-3 seconds**. The appengine dashboard is showing latencies of **1.5-2.5 seconds**.

When I switch to the `appengine` package, it is from **75-90ms**.

**Comment by okdave** [comment]
The difference between `appengine/datastore` and `cloud/datastore` is not surprising, though it might improve over time.

Currently the `cloud` version of datastore always uses a HTTP connection to contact the service, whereas the `appengine` version uses a heavily optimised-for-app-engine connection. I would use the latter version unless you really need to re-use the code outside the app engine environment.

Hopefully in the future, `cloud/datastore` could fall back to the most appropriate transport: it's probably worth filing a feature request for that.

I am interested in what performance differences (if any) you saw in using the cloud package on *compute engine* vs. *app engine*.

**Comment by ptomasroos** [comment]
@okdave Does this mean that Compute Engine and Container Engine users should be expecting terrible latency of the most fundamental storage service in the Google Platform?

**Comment by cnbuff410** [comment]
/sub

We ran into similar issue before. And didn't have good solution yet.

**Comment by ptomasroos** [comment]
@cnbuff410 sad to hear. Skipped using the service or which workaround was taken?

**Comment by cnbuff410** [comment]
@ptomasroos our service is fairly simple, so more optimizations based on memcache or UI design makes it usable.

**Comment by broady** [comment]
Hi all,

Thanks for the reports.

We'll soon be making an infrastructure change that should result in improved latency for the Cloud Datastore API. It should be live once we release `v1beta3`.

I'll update this thread once I know more. Sorry, I don't have a timeline.

**Comment by al3xdm** [comment]
Hi @broady , appreciate that it's difficult to give timelines, but is there any update you can give us? It's really a blocker for us running in production. In April another team member said that the fix would be live in the coming months ( https://groups.google.com/forum/#!topic/gcd-discuss/Pvpc5740lfI ), would be great to get this last hurdle out the way :)

**Comment by broady** [comment]
Sorry, no updates -- I'll update this thread once it goes live. I'm subscribed to our updates internally at Google.

There shouldn't be any change to the public API, so it shouldn't be tied to a milestone. We'd just be changing out the underlying transport.

**Comment by broady** [comment]
You can track the change here:
https://code-review.googlesource.com/#/c/3400/

**Comment by broady** [comment]
We're on `v1beta3` now, backed by gRPC. Performance should be much improved.
